src/app/controllers/recording_controller.py

```diff
@@ -114,25 +110,37 @@ def __recorderStateChanged(self, serviceState):
         elif self.__recorder.state == ServiceState.STOPPING:
             self.__stopOdasLive()
             self.__stopVideoProcessor()
+            
+            # waiting that the recorder is stopped.
+            while self.__recorder.state != serviceState.STOPPED:
+                time.sleep(0.0001)
 
         elif self.__recorder.state == ServiceState.STOPPED:
-            if self.__odasServer.state == ServiceState.STOPPED and self.__videoProcessor.state == ServiceState.STOPPED:
-                self.__recorder.terminate()
+            # waiting that all services are stopped.
+            while self.__odasServer.state != ServiceState.STOPPED or self.__videoProcessor.state != ServiceState.STOPPED:
```

Comment:
The Recorder is already looping on this state (See line 186 recorder.py). I am not sure if the controller should be looping and waiting on state changes. It could either be the controller or recorder that waits. We will talk about it